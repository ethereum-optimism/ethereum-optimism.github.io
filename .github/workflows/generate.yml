name: Generate Token List

on:
  workflow_dispatch:
  push:
    branches: [master]

jobs:
  generate:
    name: generate
    if: |
      github.event.head_commit.committer.username != 'github-actions[bot]'
        && github.repository == 'ethereum-optimism/ethereum-optimism.github.io'

    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v2
        with:
          token: ${{ secrets.PAT }}
          fetch-depth: 0

      - uses: pnpm/action-setup@v2
        with:
          version: 8.6.0

      - name: Use Node.js 16.x
        uses: actions/setup-node@v2
        with:
          node-version: 16.x
          cache: pnpm

      - name: Install dependencies
        run: pnpm install --frozen-lockfile

      - name: Patch token list version
        run: |
          git config --global user.email "mergify[bot]@users.noreply.github.com"
          git config --global user.name "mergify[bot]"
          npm version patch

      - name: Generate token list
        run: |
          pnpm generate:ci

      - name: Commit token list
        uses: stefanzweifel/git-auto-commit-action@v4
        with:
          commit_message: 'bot(ci): generate token list'
          file_pattern: optimism.tokenlist.json
