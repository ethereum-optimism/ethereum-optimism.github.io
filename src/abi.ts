export const FACTORY_ABI = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: 'address',
        name: '_l1Token',
        type: 'address',
      },
      {
        indexed: true,
        internalType: 'address',
        name: '_l2Token',
        type: 'address',
      },
    ],
    name: 'StandardL2TokenCreated',
    type: 'event',
  },
  {
    inputs: [
      { internalType: 'address', name: '_l1Token', type: 'address' },
      { internalType: 'string', name: '_name', type: 'string' },
      { internalType: 'string', name: '_symbol', type: 'string' },
    ],
    name: 'createStandardL2Token',
    outputs: [],
    stateMutability: 'nonpayable',
    type: 'function',
  },
]

export const TOKEN_ABI = [
  {
    constant: true,
    inputs: [],
    name: 'name',
    outputs: [
      {
        name: '',
        type: 'string',
      },
    ],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    constant: true,
    inputs: [],
    name: 'decimals',
    outputs: [
      {
        name: '',
        type: 'uint8',
      },
    ],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
  {
    constant: true,
    inputs: [],
    name: 'symbol',
    outputs: [
      {
        name: '',
        type: 'string',
      },
    ],
    payable: false,
    stateMutability: 'view',
    type: 'function',
  },
]
