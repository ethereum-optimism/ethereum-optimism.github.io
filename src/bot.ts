import crypto from 'crypto'
import fs from 'fs'
import path from 'path'

import {
  BaseServiceV2,
  ExpressRouter,
  validators,
} from '@eth-optimism/common-ts'
import { Octokit } from 'octokit'
import extract from 'extract-zip'
import uuid from 'uuid'

import { version } from '../package.json'

type TOptions = {
  secret: string
  pat: string
  tempdir: string
}

type TMetrics = {}

type TState = {
  gh: Octokit
}

export class Bot extends BaseServiceV2<TOptions, TMetrics, TState> {
  constructor(options?: Partial<TOptions>) {
    super({
      name: 'token-list-bot',
      version,
      options,
      optionsSpec: {
        secret: {
          secret: true,
          desc: 'secret used to check webhook validity',
          validator: validators.str,
        },
        pat: {
          secret: true,
          desc: 'personal access token for github',
          validator: validators.str,
        },
        tempdir: {
          desc: 'temporary directory for storing files',
          validator: validators.str,
          default: './tmp',
        },
      },
      metricsSpec: {
        // ...
      },
    })
  }

  async init(): Promise<void> {
    this.state.gh = new Octokit({
      auth: this.options.pat,
    })

    // Create the temporary directory if it doesn't exist.
    if (!fs.existsSync(this.options.tempdir)) {
      fs.mkdirSync(this.options.tempdir)
    }
  }

  async routes(router: ExpressRouter): Promise<void> {
    router.post('/webhook', async (req: any, res: any) => {
      const id = uuid.v4()
      try {
        // We'll need this later
        const owner = 'ethereum-optimism'
        const repo = 'ethereum-optimism.github.io'

        // Compute the HMAC of the request body
        const sig = Buffer.from(req.get('X-Hub-Signature-256') || '', 'utf8')
        const hmac = crypto.createHmac('sha256', this.options.secret)
        const digest = Buffer.from(
          `sha256=${hmac.update(req.rawBody).digest('hex')}`,
          'utf8'
        )

        // Check that the HMAC is valid
        if (!crypto.timingSafeEqual(digest, sig)) {
          return res
            .status(200)
            .json({ ok: false, message: 'invalid signature on workflow' })
        }

        // Make sure we're only looking at "Validate PR" workflows
        if (req.body.workflow.name !== 'Validate PR') {
          return res
            .status(200)
            .json({ ok: false, message: 'incorrect workflow' })
        }

        const artifactQueryResponse = await this.state.gh.request(
          'GET /repos/{owner}/{repo}/actions/runs/{run_id}/artifacts',
          {
            owner,
            repo,
            run_id: req.body.workflow_run.id,
          }
        )

        if (artifactQueryResponse.data.total_count !== 1) {
          return res
            .status(200)
            .json({ ok: false, message: 'incorrect number of artifacts' })
        }

        const artifact = artifactQueryResponse.data.artifacts[0]
        if (artifact.name !== 'logs-artifact') {
          return res
            .status(200)
            .json({ ok: false, message: 'incorrect artifact name' })
        }

        const artifactDownloadResponse = await this.state.gh.request(
          'GET /repos/{owner}/{repo}/actions/artifacts/{artifact_id}/{archive_format}',
          {
            owner,
            repo,
            artifact_id: artifact.id,
            archive_format: 'zip',
          }
        )

        const tempdir = path.resolve(this.options.tempdir, id)
        if (!fs.existsSync(tempdir)) {
          fs.mkdirSync(tempdir)
        }

        const temploc = path.resolve(tempdir, `${artifact.id}.zip`)
        const tempout = path.resolve(tempdir, `${artifact.id}-out`)

        fs.writeFileSync(
          temploc,
          Buffer.from(
            new Uint8Array(artifactDownloadResponse.data as ArrayBuffer)
          )
        )

        await extract(temploc, {
          dir: path.resolve(tempout),
        })

        const pr = fs.readFileSync(path.resolve(tempout, 'pr.txt'), 'utf8')
        const err = fs.readFileSync(path.resolve(tempout, 'err.txt'), 'utf8')
        const std = fs.readFileSync(path.resolve(tempout, 'std.txt'), 'utf8')

        if (err.length > 0) {
          await this.state.gh.request(
            'POST /repos/{owner}/{repo}/issues/{issue_number}/comments',
            {
              owner,
              repo,
              issue_number: parseInt(pr, 10),
              body: `Got some errors while validating this PR. You will need to fix these errors before this PR can be reviewed.\n\`\`\`\n${err}\`\`\``,
            }
          )

          return res
            .status(200)
            .json({ ok: true, message: 'ok with error comment on PR' })
        }

        const warns = std.split('\n').filter((line) => {
          return line.startsWith('warning')
        })

        if (warns.length > 0) {
          await this.state.gh.request(
            'POST /repos/{owner}/{repo}/issues/{issue_number}/comments',
            {
              owner,
              repo,
              issue_number: parseInt(pr, 10),
              body: `Got some warnings while validating this PR. This is usually OK but this PR will require manual review if you are unable to resolve these warnings.\n\`\`\`\n${warns.join(
                '\n'
              )}\n\`\`\``,
            }
          )

          await this.state.gh.request(
            'POST /repos/{owner}/{repo}/issues/{issue_number}/labels',
            {
              owner,
              repo,
              issue_number: parseInt(pr, 10),
              labels: ['requires-manual-review'],
            }
          )

          return res
            .status(200)
            .json({ ok: true, message: 'ok with warning comment on PR' })
        }

        // Can safely remove requires-manual-review if we got here without warnings! Sometimes the
        // label will be on the PR because a previous iteration of the workflow had warnings but the
        // warnings were cleared up.
        await this.state.gh.request(
          'DELETE /repos/{owner}/{repo}/issues/{issue_number}/labels/{name}',
          {
            owner,
            repo,
            issue_number: parseInt(pr, 10),
            name: 'requires-manual-review',
          }
        )

        return res.status(200).json({ ok: true, message: 'noice!' })
      } catch (err) {
        this.logger.error(err)
        return res.status(500).json({ ok: false, message: 'unexpected error' })
      } finally {
        // Always clean up the tempdir.
        const tempdir = path.resolve(this.options.tempdir, id)
        if (fs.existsSync(tempdir)) {
          fs.rmdirSync(tempdir, { recursive: true })
        }
      }
    })
  }

  async main(): Promise<void> {
    // nothing to do here
  }
}

if (require.main === module) {
  const bot = new Bot()
  bot.run()
}
