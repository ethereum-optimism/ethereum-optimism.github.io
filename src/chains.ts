import { ethers } from 'ethers'

import { Chain, L1Chain, L2Chain, Network } from './types'

const DEFAULT_INFURA_KEY = '84842078b09946638c03157f83405213'

export const NETWORK_DATA: Record<Chain, Network> = {
  ethereum: {
    id: 1,
    name: 'Mainnet',
    provider: new ethers.providers.InfuraProvider('homestead'),
    layer: 1,
  },
  optimism: {
    id: 10,
    name: 'Optimism',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://mainnet.optimism.io'
    ),
    layer: 2,
  },
  base: {
    id: 8453,
    name: 'Base',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://mainnet.base.org',
    ),
    layer: 2,
  },
  pgn: {
    id: 424,
    name: 'PGN - Public Goods Network',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://rpc.publicgoods.network'
    ),
    layer: 2,
  },
  mode: {
    id: 34443,
    name: 'Mode',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://mainnet.mode.network',
    ),
    layer: 2,
  },
  sepolia: {
    id: 11155111,
    name: 'Sepolia',
    provider: new ethers.providers.StaticJsonRpcProvider(`https://sepolia.infura.io/v3/${DEFAULT_INFURA_KEY}`, 11155111),
    layer: 1,
  },
  'optimism-sepolia': {
    id: 11155420,
    name: 'Optimism Sepolia',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://sepolia.optimism.io'
    ),
    layer: 2,
  },
  'base-sepolia': {
    id: 84532,
    name: 'Base Sepolia',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://sepolia.base.org',
      84532
    ),
    layer: 2,
  },
  'pgn-sepolia': {
    id: 58008,
    name: 'PGN Sepolia',
    provider: new ethers.providers.StaticJsonRpcProvider(
      'https://rpc.sepolia.publicgoods.network'
    ),
    layer: 2,
  },
}

interface L2BridgeInformation {
  l2StandardBridgeAddress: string
}

interface L1BridgeInformation {
  l2Chain: L2Chain
  l1StandardBridgeAddress: string
}

export const L2_STANDARD_BRIDGE_INFORMATION: Record<
  L2Chain,
  L2BridgeInformation
> = {
  optimism: {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  base: {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  pgn: {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  mode: {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  'optimism-sepolia': {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  'base-sepolia': {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
  'pgn-sepolia': {
    l2StandardBridgeAddress: '0x4200000000000000000000000000000000000010',
  },
}

export const L2_TO_L1_PAIR: Partial<Record<L2Chain, L1Chain>> = {
  optimism: 'ethereum',
  base: 'ethereum',
  pgn: 'ethereum',
  mode: 'ethereum',
  'optimism-sepolia': 'sepolia',
  'base-sepolia': 'sepolia',
  'pgn-sepolia': 'sepolia'
}

export const L1_STANDARD_BRIDGE_INFORMATION: Record<
  L1Chain,
  L1BridgeInformation[]
> = {
  ethereum: [
    {
      l2Chain: 'optimism',
      l1StandardBridgeAddress: '0x99C9fc46f92E8a1c0deC1b1747d010903E884bE1',
    },
    {
      l2Chain: 'base',
      l1StandardBridgeAddress: '0x3154Cf16ccdb4C6d922629664174b904d80F2C35',
    },
    {
      l2Chain: 'pgn',
      l1StandardBridgeAddress: '0xD0204B9527C1bA7bD765Fa5CCD9355d38338272b',
    },
    {
      l2Chain: 'mode',
      l1StandardBridgeAddress: '0x735aDBbE72226BD52e818E7181953f42E3b0FF21',
    },
  ],
  sepolia: [
    {
      l2Chain: 'optimism-sepolia',
      l1StandardBridgeAddress: '0xFBb0621E0B23b5478B630BD55a5f21f67730B0F1',
    },
    {
      l2Chain: 'pgn-sepolia',
      l1StandardBridgeAddress: '0xFaE6abCAF30D23e233AC7faF747F2fC3a5a6Bfa3',
    },
    {
      l2Chain: 'base-sepolia',
      l1StandardBridgeAddress: '0xfd0Bf71F60660E2f608ed56e1659C450eB113120',
    },
  ]
}
