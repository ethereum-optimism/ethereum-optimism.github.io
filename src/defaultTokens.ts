// All entries must match the name of a folder under the data/ directory and be upper-cased.
export const defaultTokenDataFolders = new Set([
  'ETH',
  'WETH',
  'USDC',
  'USDT',
  'SUSD',
  'LUSD',
  'DAI',
  'TUSD',
  'WSTETH',
  'CBETH',
  'LDO',
  'RETH',
  'KNC',
  'AAVE',
  'LINK',
  'WBTC',
  'POOL',
  'LDO',
  'UNI',
  'SNX',
  'THALES',
  'LYRA',
  'STKLYRA',
  'DHT',
  'PERP',
  'BAL',
  'CRV',
  'GTC',
  'ZRX',
  'WLD',
])
