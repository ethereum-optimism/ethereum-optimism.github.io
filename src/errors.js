class ErrInvalidDataJson extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidDataJson";
  }
}

class ErrInvalidLogoFile extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidLogoFile";
  }
}

class ErrInvalidTokenDecimals extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidTokenDecimals";
  }
}

class ErrInvalidTokenSymbol extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidTokenSymbol";
  }
}

class ErrInvalidTokenName extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidTokenName";
  }
}

class ErrInvalidTokenList extends Error {
  constructor(message) {
    super(message);
    this.name = "ErrInvalidTokenList";
  }
}

module.exports = {
  ErrInvalidDataJson,
  ErrInvalidLogoFile,
  ErrInvalidTokenDecimals,
  ErrInvalidTokenSymbol,
  ErrInvalidTokenName,
  ErrInvalidTokenList,
};
