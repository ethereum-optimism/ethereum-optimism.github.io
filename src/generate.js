const fs = require("fs");
const path = require("path");
const glob = require("glob");
const assert = require("assert");
const { URLSearchParams } = require("url");

const fetch = require("node-fetch");
const Validator = require("jsonschema").Validator;
const Ajv = require("ajv");
const addFormats = require("ajv-formats");
const { schema } = require("@uniswap/token-lists");
const { ethers } = require("ethers");
const { sleep } = require("@eth-optimism/core-utils");

const { version } = require("../package.json");
const { TOKEN_DATA_SCHEMA } = require("./schemas");
const { NETWORK_DATA } = require("./chains");
const errors = require("./errors");

const tokenABI = require("./abi/token.json");
const factoryABI = require("./abi/factory.json");

/**
 * Base URL where static assets are hosted.
 */
const BASE_URL = "https://ethereum-optimism.github.io";

/**
 * Call this to make CI require manual review.
 *
 * @param message Helpful message to display to the user.
 */
const review = (token, chain, message) => {
  console.log(`requires manual review: (${token} on ${chain}) ${message}`)
}

/**
 * Generates and validates the token list JSON object from the data folder.
 *
 * @param tokens List of tokens to run validation on.
 * @returns Generated and validated token list JSON object.
 */
const generate = async (tokens) => {
  const list = {
    name: "Optimism",
    logoURI: `${BASE_URL}/optimism.svg`,
    keywords: ["scaling", "layer2", "infrastructure"],
    timestamp: new Date().toISOString(),
    tokens: [],
    version: {
      major: parseInt(version.split(".")[0]),
      minor: parseInt(version.split(".")[1]),
      patch: parseInt(version.split(".")[2]),
    },
  };

  const datadir = path.resolve(__dirname, "../data/");
  const folders = fs.readdirSync(datadir).sort((a, b) => {
    return a.toLowerCase().localeCompare(b.toLowerCase());
  });

  const cgret = await fetch("https://tokens.coingecko.com/uniswap/all.json");
  const cg = await cgret.json();

  for (const folder of folders) {
    const datafile = path.join(datadir, folder, "data.json");
    assert(fs.existsSync(datafile), `${datafile} does not exist (${folder})`);
    const data = require(datafile);

    const logofiles = glob.sync(`${path.join(datadir, folder)}/logo.{png,svg}`);
    if (logofiles.length !== 1) {
      throw new errors.ErrInvalidLogoFile(
        `${folder}: exactly one logo file must be present, make sure your logo file is named either "logo.png" or "logo.svg"`
      );
    }

    if (tokens && tokens.includes(folder)) {
      console.log(`validating ${folder}`);
      const v = new Validator();
      const result = v.validate(data, TOKEN_DATA_SCHEMA);
      if (!result.valid) {
        throw new errors.ErrInvalidDataJson(
          `${folder}: data.json is not valid:\n${result.errors
          .map((err) => {
            return ` - ${err.property}: ${err.message}`;
          })
          .join("\n")}\n`
        );
      }
    }

    for (const [chain, token] of Object.entries(data.tokens)) {
      if (tokens && tokens.includes(folder)) {
        console.log(`validating ${folder} on chain ${chain}`);

        // Check to make sure the website will load
        for (let i = 0; i < 5; i++) {
          try {
            await fetch(data.website);
            break;
          } catch (err) {
            if (i < 4) {
              console.log(`website did not load, trying again in 5s...`)
              await sleep(5000);
            } else {
              throw new errors.ErrWebsiteLoadFailed(
                `${folder}: website did not load: ${data.website}`
              );
            }
          }
        }

        if (folder !== "ETH" && data.nonstandard !== true) {
          const contract = new ethers.Contract(
            token.address,
            tokenABI,
            NETWORK_DATA[chain].provider
          );

          // Check that the token exists on this chain
          if (
            await contract.provider.getCode(token.address) === "0x"
          ) {
            throw new errors.ErrNoTokenCode(
              `${chain} ${folder} token has no code at address`
            );
          }

          // Check that the token has the correct decimals
          if (token.overrides?.decimals === undefined) {
            if (data.decimals !== (await contract.decimals())) {
              throw new errors.ErrInvalidTokenDecimals(
                `${chain} ${folder} decimals does not match data.json decimals`
              );
            }
          } else {
            review(folder, chain, "has decimal override");
          }

          // Check that the token has the correct symbol
          if (token.overrides?.symbol === undefined) {
            if (data.symbol !== (await contract.symbol())) {
              throw new errors.ErrInvalidTokenSymbol(
                `${chain} ${folder} symbol does not match data.json symbol`
              );
            }
          } else {
            review(folder, chain, "has symbol override");
          }

          // Check that the token has the correct name
          if (token.overrides?.name === undefined) {
            if (data.name !== (await contract.name())) {
              throw new errors.ErrInvalidTokenName(
                `${chain} ${folder} name does not match data.json name`
              );
            }
          } else {
            review(folder, chain, "has name override");
          }

          if (chain === "ethereum") {
            const found = cg.tokens.find((t) => {
              return t.address.toLowerCase() === token.address.toLowerCase();
            })

            // Trigger manual review if the Ethereum token is not in the CG token list
            if (!found) {
              review(folder, chain, "not found on CoinGecko token list");
            }
          }

          if (chain.startsWith('optimism')) {
            const factory = new ethers.Contract(
              '0x4200000000000000000000000000000000000012',
              factoryABI,
              NETWORK_DATA[chain].provider
            );

            const events = await factory.queryFilter(
              factory.filters.StandardL2TokenCreated(undefined, token.address)
            );

            // Trigger review if not created by standard token factory.
            if (events.length === 0) {
              review(folder, chain, "not created by standard token factory");
            }
          } else {
            // Make sure the token is verified on Etherscan.
            // Etherscan API is heavily rate limited, so sleep for 1s to avoid errors.
            await sleep(1000);
            const { result } = await (
              await fetch(`https://api${chain === 'ethereum' ? "" : `-${chain}`}.etherscan.io/api?` + new URLSearchParams({
                module: "contract",
                action: "getsourcecode",
                address: token.address,
                apikey: process.env.ETHERSCAN_API_KEY
              }))
            ).json()

            // Trigger review if code not verified on Etherscan
            if (result[0].ABI === "Contract source code not verified") {
              review(folder, chain, "code not verified on Etherscan");
            }
          }
        } else {
          review(folder, chain, "nonstandard token");
        }
      }

      list.tokens.push({
        chainId: NETWORK_DATA[chain].id,
        address: token.address,
        name: token.overrides?.name ?? data.name,
        symbol: token.overrides?.symbol ?? data.symbol,
        decimals: token.overrides?.decimals ?? data.decimals,
        logoURI: `${BASE_URL}/data/${folder}/logo.${
          logofiles[0].endsWith("png") ? "png" : "svg"
        }`,
        extensions: {
          optimismBridgeAddress:
            token.overrides?.bridge ?? NETWORK_DATA[chain].bridge,
        },
      });
    }
  }

  // Verify that the final generated token list is valid
  const ajv = new Ajv({ allErrors: true, verbose: true });
  addFormats(ajv);
  const validator = ajv.compile(schema);
  if (!validator(list)) {
    throw new errors.ErrInvalidTokenList(
      `final token list is not valid:\n ${validator.errors}`
    );
  }

  return list;
};

module.exports = {
  generate,
};
