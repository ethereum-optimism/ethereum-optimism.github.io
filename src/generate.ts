import fs from 'fs'
import path from 'path'

import { glob } from 'glob'

import { version } from '../package.json'
import {
  L1_STANDARD_BRIDGE_INFORMATION,
  L2_STANDARD_BRIDGE_INFORMATION,
  NETWORK_DATA,
} from './chains'
import { L2Chain, Token, TokenData, isL1Chain, isL2Chain } from './types'
import { defaultTokenDataFolders } from './defaultTokens'

/**
 * Base URL where static assets are hosted.
 */
const BASE_URL = 'https://ethereum-optimism.github.io'

/**
 * Generates a token list from the data in the data folder.
 *
 * @param datadir Directory containing data files.
 *
 * @returns Generated token list JSON object.
 */
export const generate = (datadir: string) => {
  return fs
    .readdirSync(datadir)
    .sort((a, b) => {
      return a.toLowerCase().localeCompare(b.toLowerCase())
    })
    .map((folder) => {
      const data: TokenData = JSON.parse(
        fs.readFileSync(path.join(datadir, folder, 'data.json'), 'utf8')
      )
      const logofiles = glob.sync(
        `${path.join(datadir, folder)}/logo.{png,svg}`
      )
      const logoext = logofiles[0].endsWith('png') ? 'png' : 'svg'
      return Object.entries(data.tokens).map(([chain, token]) => {
        const bridges = !data.nobridge
          ? Object.assign({}, ...getBridges(data, chain, token))
          : {}
        const out = {
          chainId: NETWORK_DATA[chain].id,
          address: token.address,
          name: token.overrides?.name ?? data.name,
          symbol: token.overrides?.symbol ?? data.symbol,
          decimals: token.overrides?.decimals ?? data.decimals,
          logoURI: `${BASE_URL}/data/${folder}/logo.${logoext}`,
          extensions: {
            ...bridges,
            opListId: defaultTokenDataFolders.has(folder.toUpperCase())
              ? 'default'
              : 'extended',
            opTokenId: folder,
          },
        }
        return out
      })
    })
    .reduce(
      (list, tokens) => {
        list.tokens = list.tokens.concat(tokens)
        return list
      },
      {
        name: 'Superchain Token List',
        logoURI: `${BASE_URL}/optimism.svg`,
        keywords: ['scaling', 'layer2', 'infrastructure'],
        timestamp: new Date().toISOString(),
        tokens: [],
        version: {
          major: parseInt(version.split('.')[0], 10),
          minor: parseInt(version.split('.')[1], 10),
          patch: parseInt(version.split('.')[2], 10),
        },
      }
    )
}

const getBridges = (tokenData: TokenData, chain: string, token: Token) => {
  if (isL2Chain(chain)) {
    const tokenBridgeOverride = token.overrides?.bridge
    if (tokenBridgeOverride && typeof tokenBridgeOverride !== 'string') {
      throw new Error('L2 Bridge override should be a string')
    }
    const networkSep = chain.indexOf('-')
    const chainName = networkSep === -1 ? chain : chain.slice(0, networkSep)
    const bridgeKey = `${chainName}BridgeAddress`
    return [
      {
        [bridgeKey]:
          tokenBridgeOverride ??
          L2_STANDARD_BRIDGE_INFORMATION[chain].l2StandardBridgeAddress,
      },
    ]
  }

  if (isL1Chain(chain)) {
    const l2ChainsForL1 = L1_STANDARD_BRIDGE_INFORMATION[chain].map(
      (l1Bridge) => l1Bridge.l2Chain
    )
    const l2ChainsSupported = Object.entries(tokenData.tokens)
      .filter(
        ([tokenChain]) =>
          isL2Chain(tokenChain) && l2ChainsForL1.includes(tokenChain)
      )
      .map(([l2Chain]) => l2Chain as L2Chain)
    return l2ChainsSupported.map((l2Chain) => {
      const l1StandardBridgeInfoForL2 = L1_STANDARD_BRIDGE_INFORMATION[
        chain
      ].find((l1BridgeInfo) => l1BridgeInfo.l2Chain === l2Chain)
      const tokenBridgeOverride = token.overrides?.bridge
      if (tokenBridgeOverride && typeof tokenBridgeOverride === 'string') {
        throw new Error(
          'L1 Bridge override should be a map from l2 chain to bridge address'
        )
      }
      const networkSep = l2Chain.indexOf('-')
      const chainName = networkSep === -1 ? l2Chain : l2Chain.slice(0, networkSep)
      const bridgeKey = `${chainName}BridgeAddress`
      return {
        [bridgeKey]:
          tokenBridgeOverride?.[l2Chain] ??
          l1StandardBridgeInfoForL2.l1StandardBridgeAddress,
      }
    })
  }

  throw new Error('unsupported chain')
}
