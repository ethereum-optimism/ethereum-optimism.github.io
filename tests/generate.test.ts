import path from 'path'

import { generate } from '../src/generate'

test("'generate' script parse data dir and compile correct token list", async () => {
  jest
    .spyOn(path, 'resolve')
    .mockReturnValueOnce(path.resolve(__dirname, 'data'))

  const mockDate = new Date(1660755600000)
  jest.spyOn(global, 'Date').mockImplementation(() => mockDate as any)

  const tokenList = generate(path.resolve(__dirname, 'data'))
  expect(tokenList).toMatchSnapshot({
    timestamp: mockDate.toISOString(),
    version: {
      patch: expect.any(Number),
    },
  })
})
