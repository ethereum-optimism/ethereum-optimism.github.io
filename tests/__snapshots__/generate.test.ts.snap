// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'generate' script parse data dir and compile correct token list 1`] = `
Object {
  "keywords": Array [
    "scaling",
    "layer2",
    "infrastructure",
  ],
  "logoURI": "https://ethereum-optimism.github.io/optimism.svg",
  "name": "Optimism",
  "timestamp": "2022-08-17T17:00:00.000Z",
  "tokens": Array [
    Object {
      "address": "0x0000000000000000000000000000000000000000",
      "chainId": 1,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x99C9fc46f92E8a1c0deC1b1747d010903E884bE1",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/ETH/logo.svg",
      "name": "Ether",
      "symbol": "ETH",
    },
    Object {
      "address": "0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000",
      "chainId": 10,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x4200000000000000000000000000000000000010",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/ETH/logo.svg",
      "name": "Ether",
      "symbol": "ETH",
    },
    Object {
      "address": "0x0000000000000000000000000000000000000000",
      "chainId": 42,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x22F24361D548e5FaAfb36d1437839f080363982B",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/ETH/logo.svg",
      "name": "Ether",
      "symbol": "ETH",
    },
    Object {
      "address": "0xDeadDeAddeAddEAddeadDEaDDEAdDeaDDeAD0000",
      "chainId": 69,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x4200000000000000000000000000000000000010",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/ETH/logo.svg",
      "name": "Ether",
      "symbol": "ETH",
    },
    Object {
      "address": "0x4200000000000000000000000000000000000042",
      "chainId": 10,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x4200000000000000000000000000000000000010",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/OP/logo.png",
      "name": "Optimism",
      "symbol": "OP",
    },
    Object {
      "address": "0x4200000000000000000000000000000000000042",
      "chainId": 69,
      "decimals": 18,
      "extensions": Object {
        "optimismBridgeAddress": "0x4200000000000000000000000000000000000010",
      },
      "logoURI": "https://ethereum-optimism.github.io/data/OP/logo.png",
      "name": "Optimism",
      "symbol": "OP",
    },
  ],
  "version": Object {
    "major": 4,
    "minor": 3,
    "patch": Any<Number>,
  },
}
`;
